### PR TITLE
feat(SQS): ingest cwPools only when modified rather than every block

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #XXX

## What is the purpose of the change

As a workaround, we currently process every cwPool on every block. This was done because, the pool object chain side only changes when one of two things happen:
1. Pool is created
2. Pool is migrated

This is unlike all other pool types, where the pool object chain side is changed on every swap. We therefore do the following:

1. Ingest all cwPools on cold start, mapping the pool address to the pool object (which only changes in the two cases described above)
2. We listen to onWrite on the cosmwasmpool store, so that if one of the two cases described above occurs, we modify the pool address to pool struct mapping
3. We now also listen to onWrite on the bank store. We check each balance change address against the pool address to pool struct mapping. If the pool address has it's balance change, we add the stored pool struct (which doesn't change) to the change cwPools struct for this block, to be ingested by SQS

## Testing and Verifying

I set up a node against mainnet, and added prints to ensure that cwPools were now being detected as changed when swaps occur. 

I recommend this be tested in stage as well. I wasn't able to test against mainnet pool creations / pool migrations, but should theoretically work as described above.

## Summary by CodeRabbit

- **New Features**
  - Enhanced tracking and management of CosmWasm pools for improved data processing efficiency.
  - Introduced functionalities to track changes and balance updates in CosmWasm pools accurately.

- **Bug Fixes**
  - Corrected the detection of CW pool store updates, preventing unnecessary data processing per block.

- **Refactor**
  - Improved error handling and internal logic in block processing methods for enhanced stability and performance.

- **Tests**
  - Implemented new tests for balance write functionalities related to CosmWasm pools to ensure reliability.

### CODE REVIEW

File: ingest/sqs/ingester.go [hunk 1]
```diff
@@ -51,6 +52,15 @@ func (i *sqsIngester) ProcessAllBlockData(ctx sdk.Context) error {
...
 	}
 
+	// Call the callback function for each cwPool so we can create the initial cwPool address to pool object mapping
+	// This then gets modified as cwPools are created and/or migrated
+
+	for _, cwPool := range cosmWasmPools {
+		if onCosmWasmPool != nil {
+			onCosmWasmPool(cwPool)
+		}
+	}
```
Comment (p0mvn) [inline ingest/sqs/ingester.go:62]: Can we instead return the cosmwasm pools and let the caller iterate over them?

**Review by p0mvn — APPROVED** [review]
Great work.

Let's work on a stage deployment and ask QA to validate this post-merge 🙏

File: ingest/sqs/ingester.go (hunk 1 above)
Comment (czarcas7ic) [reply ingest/sqs/ingester.go:62]: Lol I actually had it like that but this felt even cleaner, I can revert to how I originally had it though

Comment (p0mvn) [reply ingest/sqs/ingester.go:62]: While the trade-off is negligible today, I think that callbacks are more difficult to maintain and keep testable as the system evolves long-term

**Pull request merged**